### ISSUE TITLE
NVFLARE per-class metrics silently dropped under burst — DB stores a random 1..N of N per-class values per round

### ISSUE THREAD

**Issue by atriaybagur** [londonaicentre/FLIP#739]
## Summary

NVFLARE per-class training/validation/test metrics are **silently dropped in transit**: the app emits a fixed N values per round (one per class) but the `fl_metrics` table stores a random **1..N** of them. Metrics are sent as **one federated NVFLARE event per scalar**, so a training round fires a burst (~30 events/epoch for a 5-class head × {precision,recall,f1} × {train,val}; 15 at once in the validator). Federated events are best-effort, and a large fraction of the burst never reaches storage. The single once-per-round loss metric is never affected.

This is the actual cause of the "different number of datapoints per round" seen on the metrics chart — it is **real missing rows**, not a plotting artifact, and it is separate from the label-collapse issue (#737 / #738).

## Evidence (prod model `d7d65959-3e40-4744-b853-0ece208df840`)

Point counts per round per trust, straight from `GET /api/model/{id}/metrics` (one JSON point = one `fl_metrics` row):

```
[TRAIN_LOSS]  BDMS: r1=1 r2=1 … r100=1        # 1 event/round  → never dropped
[TRAIN_LOSS]  GSTT: r1=1 r2=1 … r100=1
[VAL-RECALL]  GSTT: r1=4 r2=3 r3=3 r4=2 r5=2 r6=1 …  r55=5 r60=5 …   # 5 emitted/round → 1..5 stored
[…-F1-SCORE]  GSTT: … r54=3 r55=5 r56=4 r60=2 r61=4 r62=3 r63=5 …
[TEST-F1-SCORE]  GSTT: r0=2      # validator emits 5 (one/class); only 2 stored
[TEST-PRECISION] GSTT: r0=1      # 5 emitted → 1 stored
[TEST-RECALL]    GSTT: r0=2      # 5 emitted → 2 stored
```

The correlation is unambiguous: **1 event/round ⇒ 0% loss; a burst of many ⇒ heavy, random loss** (the validator burst lost ~10 of 15).

## Why this is transport loss, not app/skip/NaN behaviour

The apps emit a **fixed N unconditionally** — undefined/NaN per-class metrics are coerced to `0.0` and still sent, never skipped:

- `fl-tutorials/nvflare/image_classification/xray_classification/app_files/trainer.py` (NaN→0.0 then unconditional `send_metrics_value` per lesion) and the same in the Ark+ fine-tuning app that produced this model.
- Response contains no `NaN` (the fetch's `JSON.parse` succeeds), so missing points are genuinely **absent rows**, not dropped-NaN.

So N are emitted and <N are stored ⇒ loss is between the client `send_metrics_value` call and the `fl_metrics` insert.

## Where it happens

One event per scalar, all best-effort:

- `flip-utils/flip/nvflare/metrics.py:22-71` — `send_metrics_value` fires one `FlipEvents.SEND_RESULT` **federated** event per call.
- `flip-utils/flip/nvflare/controllers/scatter_and_gather.py:123-136` — server `handle_event` relays one POST per received event (`handle_metrics_event`).
- `flip-utils/flip/nvflare/metrics.py:74-120` — `handle_metrics_event` → one `flip.send_metrics` POST.

The client→server federated-event relay (NVFLARE aux channel) has no application-level ack/retry, so burst events are lost there (candidate: relay best-effort delivery; secondary candidate: shared `FLContextKey.EVENT_DATA` prop clobbered when events are processed back-to-back). The Client-API path has the same shape via `FlipAnalyticsBridge` (`SummaryWriter.add_scalar` → one analytic event per scalar), so it is likely affected too.

**Flower is not affected**: clients return all of a round's metrics in one reply `Message`/`MetricRecord` (`flip-utils/flip/flower/metrics.py`), which is delivered reliably; only the server→hub POSTs are per-metric and run sequentially server-side.

## Proposed fix

Stop firing one federated event per scalar. Batch a round's metrics into a **single** event / payload:

1. Add a batch send (e.g. `send_metrics_values(round, {label: value, …})`) that fires **one** `SEND_RESULT` carrying all of the round's `(label, value)` pairs; have `handle_metrics_event` forward them in one POST (extend the `/model/{id}/metrics` endpoint to accept a list). One aux relay + one insert per round → no burst, atomic per round.
2. Migrate the app templates to the batch call (trainer + validator, both backends' shared surface).
3. Optionally, make the relay delivery reliable (acked) as defence-in-depth.

## Relationship to #737 / #738

- #738 (per-class labels) is still correct and needed, but it only **disambiguates** the classes and makes a dropped class show as a per-class gap — it does **not** stop the loss.
- The "varying visible count" originally attributed on #737 to render-overlap is actually this data loss; see the correction comment there.

## Verification

- [ ] Re-run the count query after the fix: `SELECT global_round, trust, label, count(*) FROM fl_metrics WHERE model_id = '…' AND label LIKE '%F1-SCORE%' GROUP BY 1,2,3 ORDER BY 1,2;` — expect a **constant** `count(*)` (= number of classes) every round.
- [ ] Confirm the Client-API path (`FlipAnalyticsBridge`) is covered by the batch fix.
